### PR TITLE
[MIG] paydays > account_payment_term_multi_day

### PR DESCRIPTION
**Pull request by pedrobaeza**
Migrated to the new API.

### CODE REVIEW

**Comment by oihane** [comment]
:+1:

File: paydays/models/__init__.py [hunk 1]
```diff
@@ -19,4 +19,4 @@
 #
 ##############################################################################
```
Comment (bwrsandman) [inline paydays/models/__init__.py]: File fails pep8

Comment (pedrobaeza) [inline paydays/models/__init__.py]: Sorry, a little mistake. Done.

**Comment by pedrobaeza** [comment]
Can you please review this PR to merge it ASAP?

File: paydays/models/account_payment_term.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    def _check_payment_days(self):
+        error = False
+        try:
+            for day in self._decode_payment_days(self.payment_days):
+                if day <= 0 or day > 31:
+                    # To force to go to except block
+                    raise Exception()
+        except:
+            error = True
+        if error:
```
Comment (gurneyalex) [inline paydays/models/account_payment_term.py]: please change this code to:

```
payment_days = self._decode_payment_days(self.payment_days)
error = any(day <= 0 or day > 31 for day in  payment_days)
if error:
```

File: paydays/models/account_payment_term.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            raise exceptions.Warning(
+                _('Error!'),
```
Comment (gurneyalex) [inline paydays/models/account_payment_term.py]: remove exclamation mark

File: paydays/models/account_payment_term.py (hunk 2 above)
Comment (gurneyalex) [inline paydays/models/account_payment_term.py]: sorry, I see that the try..except block is also here to guard against exception in `_decode_payment_days` (typically by `int()`). 

The error handling in the code is still convoluted, and not very helpful as the error message does not give any information about what is wrong tin the field...

File: paydays/models/account_payment_term.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                if date.day <= day:
+                    if day > days_in_month:
+                        day = days_in_month
+                    new_date = date + relativedelta(day=day)
+                    break
+            if not new_date:
+                day = days[0]
+                if day > days_in_month:
+                    day = days_in_month
+                new_date = date + relativedelta(day=day, months=1)
```
Comment (gurneyalex) [inline paydays/models/account_payment_term.py]: is there not some code in python-dateutil handling this? Dateutil is a standard Odoo dependency.

**Comment by gurneyalex** [comment]
@pedrobaeza some comments above. You'll  get bonus points if you supply some automated tests :-)

File: paydays/models/account_payment_term.py (hunk 4 above)
Comment (pedrobaeza) [inline paydays/models/account_payment_term.py]: relativedelta is from dateutil: 

```
from dateutil.relativedelta import relativedelta
```

**Comment by pedrobaeza** [comment]
Hi, @gurneyalex, I have changed PR according to your comments. Thanks for your suggestions. And I have tried to add a test, but I don't get it working (I'm very new in tests). The question is that test invoices 2 and 3 are expected to be modified by this module, but the due date is being the same in all the 3 samples. Can you help me to see what's the problem?

**Comment by bwrsandman** [comment]
PR needs a rebase to be merged and tested

File: paydays/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
+    'name': "Payment days",
+    'version': "1.0",
+    'author': "NaN·tic",
```
Comment (gurneyalex) [inline paydays/__openerp__.py]: please add "Odoo Community Association (OCA)"

**Comment by omar7r** [comment]
:+1: tested

**Comment by bwrsandman** [comment]
Needs a rebase

**Comment by pedrobaeza** [comment]
I have rebased and change the name and functionality of the module, moving payment days to lines, so please review again.

**Comment by oihane** [comment]
@pedrobaeza travis continues failing, I think you have to update the travis

File: account_payment_term_multi_day/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    'name': "Multiple payment days for payment terms",
+    'version': "1.0",
+    'author': "NaN·tic, "
+              "Serv. Tecnol. Avanzados - Pedro M. Baeza, "
+              "Odoo Community Association (OCA)",
+    'category': 'Accounting',
+    'contributors': [
+        'Pedro M. Baeza <pedro.baeza@serviciosbaeza.com>'
+    ],
+    'license': "GPL-3",
```
Comment (omar7r) [inline account_payment_term_multi_day/__openerp__.py]: AGPL?

**Comment by omar7r** [comment]
tested :+1:

**Comment by pedrobaeza** [comment]
@gurneyalex, I still need help with the failing test. Can you please check it?

File: account_payment_term_multi_day/models/account_payment_term.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+class AccountPaymentTerm(models.Model):
+    _inherit = "account.payment.term"
+
+    @api.multi
+    def compute(self, value, date_ref=False):
+        result = super(AccountPaymentTerm, self).compute(value, date_ref)[0]
+        i = -1
+        for line in self.line_ids:
+            i += 1
```
Comment (gurneyalex) [inline account_payment_term_multi_day/models/account_payment_term.py]: please use `for i, line in enumerate(self.line_ids):`

File: account_payment_term_multi_day/models/account_payment_term.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        return [result]
+
+
+class AccountPaymentTermLine(models.Model):
+    _inherit = "account.payment.term.line"
+
+    def _decode_payment_days(self, days_char):
+        # Admit space, dash and comma as separators
+        days_char = days_char.replace(' ', '-').replace(',', '-')
+        days_char = [x.strip() for x in days_char.split('-') if x]
```
Comment (gurneyalex) [inline account_payment_term_multi_day/models/account_payment_term.py:62]: if you replace the '-' and ',' with ' ' in there, you can use `days_char = days_char.split()` and not worry about stripping the strings

Comment (pedrobaeza) [inline account_payment_term_multi_day/models/account_payment_term.py:62]: But maybe you have put the days without spaces (like `5,10`), and you'll get with your proposal `510`, so it's not correct.

Comment (bwrsandman) [inline account_payment_term_multi_day/models/account_payment_term.py:62]: Add those cases to a unittest.

**Comment by pedrobaeza** [comment]
Tests changed to Python and I get a green. With one more approval I merge because it has been a long journey until this.

**Comment by Daniel-CA** [comment]
:+1:

**Pull request merged**